state-chain/pallets/cf-elections/src/electoral_systems/liveness.rs

```diff
@@ -0,0 +1,183 @@
+use crate::{
+	electoral_system::{
+		AuthorityVoteOf, ConsensusVote, ConsensusVotes, ElectionIdentifierOf, ElectionReadAccess,
+		ElectionWriteAccess, ElectoralSystem, ElectoralWriteAccess, VotePropertiesOf,
+	},
+	vote_storage::{self, VoteStorage},
+	CorruptStorageError,
+};
+use cf_primitives::AuthorityCount;
+use cf_utilities::success_threshold_from_share_count;
+use frame_support::{
+	pallet_prelude::{MaybeSerializeDeserialize, Member},
+	Parameter,
+};
+
+use itertools::Itertools;
+use sp_std::{collections::btree_map::BTreeMap, vec::Vec};
+
+pub struct Liveness<ChainBlockNumber, ChainBlockHash, BlockNumber, Hook, ValidatorId> {
+	_phantom: core::marker::PhantomData<(
+		ChainBlockNumber,
+		ChainBlockHash,
+		BlockNumber,
+		Hook,
+		ValidatorId,
+	)>,
+}
+
+pub trait OnCheckComplete<ValidatorId> {
+	fn on_check_complete(validator_ids: Vec<ValidatorId>);
+}
+
+impl<
+		ChainBlockNumber: Member + Parameter + Eq + From<u64> + Into<u64> + Copy,
+		ChainBlockHash: Member + Parameter + Eq + Ord,
+		BlockNumber: Member
+			+ Parameter
+			+ Eq
+			+ MaybeSerializeDeserialize
+			+ frame_support::sp_runtime::Saturating
+			+ Ord
+			+ Copy,
+		Hook: OnCheckComplete<ValidatorId> + 'static,
+		ValidatorId: Member + Parameter + Ord + MaybeSerializeDeserialize,
+	> ElectoralSystem for Liveness<ChainBlockNumber, ChainBlockHash, BlockNumber, Hook, ValidatorId>
+{
+	type ValidatorId = ValidatorId;
+	type ElectoralUnsynchronisedState = ();
+	type ElectoralUnsynchronisedStateMapKey = ();
+	type ElectoralUnsynchronisedStateMapValue = ();
+
+	type ElectoralUnsynchronisedSettings = ();
+	// How many SC blocks to wait before we consider the election complete.
+	type ElectoralSettings = BlockNumber;
+	type ElectionIdentifierExtra = ();
+
+	// The external chain block number that the engines will get the hash for.
+	type ElectionProperties = ChainBlockNumber;
+
+	// The SC block number that we started the election at.
+	type ElectionState = BlockNumber;
+	type Vote = vote_storage::bitmap::Bitmap<ChainBlockHash>;
+	type Consensus = Vec<Self::ValidatorId>;
```

Comment:
Does it make sense to use a BTreeSet to make it easier to convince ourselves there are no duplicate report? (And would allow us to simply assert equality in the tests instead of having to sort etc.)